src/Pgvector.EntityFrameworkCore/VectorFunctionTranslatorPlugin.cs

```diff
@@ -0,0 +1,89 @@
+using Microsoft.EntityFrameworkCore.Storage;
+using Npgsql.EntityFrameworkCore.PostgreSQL.Query.Expressions.Internal;
+using System.Reflection;
+
+namespace Pgvector.EntityFrameworkCore;
+
+public class VectorFunctionTranslatorPlugin : IMethodCallTranslatorPlugin
+{
+    public VectorFunctionTranslatorPlugin(ISqlExpressionFactory sqlExpressionFactory, ITypeMappingSource typeMappingSource)
+    {
+        Translators = new[]
+        {
+            new VectorFunctionTranslator(sqlExpressionFactory, typeMappingSource),
+        };
+    }
+
+    public virtual IEnumerable<IMethodCallTranslator> Translators { get; }
+
+    private class VectorFunctionTranslator : IMethodCallTranslator
+    {
+        private readonly ISqlExpressionFactory _sqlExpressionFactory;
+        private readonly ITypeMappingSource _typeMappingSource;
+
+        private static readonly MethodInfo _methodL2Distance = typeof(VectorExtensions)
+            .GetRuntimeMethod(nameof(VectorExtensions.L2Distance), new[]
+            {
+                typeof(Vector),
+                typeof(Vector),
+            })!;
+
+        private static readonly MethodInfo _methodCosineDistance = typeof(VectorExtensions)
+            .GetRuntimeMethod(nameof(VectorExtensions.CosineDistance), new[]
+            {
+                typeof(Vector),
+                typeof(Vector),
+            })!;
+
+        private static readonly MethodInfo _methodMaxInnerProduct = typeof(VectorExtensions)
+            .GetRuntimeMethod(nameof(VectorExtensions.MaxInnerProduct), new[]
+            {
+                typeof(Vector),
+                typeof(Vector),
+            })!;
+
+        public VectorFunctionTranslator(ISqlExpressionFactory sqlExpressionFactory, ITypeMappingSource typeMappingSource)
+        {
+            _sqlExpressionFactory = sqlExpressionFactory;
+            _typeMappingSource = typeMappingSource;
+        }
+
+#pragma warning disable EF1001
+        public SqlExpression? Translate(
+            SqlExpression? instance,
+            MethodInfo method,
+            IReadOnlyList<SqlExpression> arguments,
+            IDiagnosticsLogger<DbLoggerCategory.Query> logger
+        )
+        {
+            var vectorOperator = method switch
+            {
+                _ when ReferenceEquals(method, _methodL2Distance) => "<->",
+                _ when ReferenceEquals(method, _methodCosineDistance) => "<=>",
+                _ when ReferenceEquals(method, _methodMaxInnerProduct) => "<#>",
+                _ => null
+            };
+
+            if (vectorOperator != null)
+            {
+                var left = arguments[0];
+                var right = arguments[1];
+
+                return new PostgresUnknownBinaryExpression(
+                    _sqlExpressionFactory.ApplyDefaultTypeMapping(left),
+                    _sqlExpressionFactory.ApplyDefaultTypeMapping(right),
+                    vectorOperator,
+                    typeof(double),
+                    left.TypeMapping
```

Comment:
This doesn't look right - shouldn't the result returned by these functions have a different type mapping as its arguments? The arguments are a vector, whereas the result is a double?

Reply:
All of these functions return `double`:
- `cosine_distance(vector, vector) → double precision`
- `inner_product(vector, vector) → double precision`
- `l2_distance(vector, vector) → double precision`

But I made it differently to avoid hard-coding of return type:
```
var resultTypeMapping = _typeMappingSource.FindMapping(method.ReturnType)!;

return new PostgresUnknownBinaryExpression(
    left: _sqlExpressionFactory.ApplyDefaultTypeMapping(left),
    right: _sqlExpressionFactory.ApplyDefaultTypeMapping(right),
    binaryOperator: vectorOperator,
    type: resultTypeMapping.ClrType,
    typeMapping: resultTypeMapping
);
```

Could this to be the right solution?

Reply:
Sure. There's no particular need to avoid hard-coding the return type - the method's return type (double) doesn't typically ever change. The problem previously was the return **type mapping** which was left.TypeMapping (vector), which wasn't right.